Backend/services/notificationService.js

```diff
@@ -132,4 +161,177 @@ const updateImportantNotificationTypes = async () => {
+      const randomHour = hours[Math.floor(Math.random() * hours.length)];
+      if (!centroids.includes(randomHour)) {
+        centroids.push(randomHour);
+      }
+    }
+
+    // Helper function to calculate distance between two points (hours)
+    const calculateDistance = (point, centroid) => Math.abs(point - centroid);
+
+    // Helper function to update centroids
+    const updateCentroids = (clusters) => {
+      return clusters.map((cluster) => {
+        if (cluster.length === 0) return 0; // Avoid division by zero
+        const sum = cluster.reduce((acc, point) => acc + point, 0);
+        return Math.round(sum / cluster.length);
+      });
+    };
+
+    let clusters = [];
+    let iterations = 0;
+
+    while (iterations < maxIterations) {
+      // Step 4: Assign each point to the nearest centroid
+      clusters = Array.from({ length: k }, () => []);
+      hours.forEach((hour) => {
+        let minDistance = Infinity;
+        let closestCentroid = 0;
+        centroids.forEach((centroid, idx) => {
+          const distance = calculateDistance(hour, centroid);
+          if (distance < minDistance) {
+            minDistance = distance;
+            closestCentroid = idx;
+          }
+        });
+        clusters[closestCentroid].push(hour);
+      });
+
+      // Step 5: Update centroids based on the mean of assigned points
+      const newCentroids = updateCentroids(clusters);
+
+      // Check for convergence if centroids do not change
+      if (JSON.stringify(newCentroids) === JSON.stringify(centroids)) break;
+
+      centroids = newCentroids;
+      iterations++;
+    }
+
+    // Step 6: Determine the best hour from the largest cluster
+    const bestCluster = clusters.reduce(
+      (maxCluster, cluster) =>
+        cluster.length > maxCluster.length ? cluster : maxCluster,
+      clusters[0]
+    );
+    const bestHour = Math.round(
+      bestCluster.reduce((acc, hour) => acc + hour, 0) / bestCluster.length
+    );
+    await prisma.userActivePeriods.upsert({
+      where: { userId: userId },
+      update: { mostActivePeriods: { set: [bestHour] } },
+      create: {
+        userId: userId,
+        mostActivePeriods: { set: [bestHour] },
+      },
+    });
+    return bestHour;
+  } catch (error) {
+    console.error("Error in K-means clustering:", error);
+  }
+};
+
+const scheduleNotifications = async (userId, newNotification, io) => {
+  const userSockets = Array.from(io.sockets.sockets.values()).filter(
+    (socket) => socket.userId == userId
+  );
+  // Add the new notification to the pending list
+  if (!pendingNotifications.has(userId)) {
+    pendingNotifications.set(userId, []);
+  }
+  pendingNotifications.get(userId).push(newNotification);
+
+  // Check if the job is already scheduled
+  if (scheduledJobs.has(userId)) {
+    return;
+  }
+
+  const bestHour = await performKMeansClustering(userId);
+  const now = new Date();
+
+  const testRunTime = new Date(now.getTime() + 10 * 1000);
```

Comment:
what is this doing? could you leave a comment above this about why you're using 10 and 1000?

Reply:
oh i see you remove this in the next commit, never mind!